### ISSUE TITLE
Material Web Components for Flask

### ISSUE THREAD

**Issue by cubetastic33** [material-components/material-components-web#2774]
## Feature Requests
I wanted to use Material Web Components after learning about it from Google I/O 18, and they said integrating these components was going to be really easy. I looked around, and it looks like you need to use node js to be able to use this, because you need to compile es2015, and also these docs here are using node js. My web app, however, uses Flask. I am wondering how it is possible to start using Material Web Components in my web app too, and I don't want to use both Flask and Node JS at the same time, after reading this SO thread: https://stackoverflow.com/questions/11158686/python-flask-alongside-with-node-js.

**Comment by touficbatache** [comment]
You can get the compiled js and css at: http://unpkg.com/material-components-web@latest/dist/

**Comment by rishabh3112** [comment]
Create templates using npm and node js and use bundled css and js files with html files

**Comment by cubetastic33** [comment]
@touficbatache yes, but the docs say that it is recommended to use npm over those link... Plus, at google io18 they said they would love to hear from us if we had other frameworks we would like to have Material Components support in.

**Comment by kfranqueiro** [comment]
The most significant benefit of including the components within your own toolchain is that it grants you access to use MDC Web's many Sass mixins, which provide a great deal of convenience and flexibility for customizing the appearance of components.

It should be feasible to use the precompiled JavaScript, but then rely on the unbuilt Sass.

The JavaScript can be referenced either via unpkg or via npm packages installed locally - the precompiled JS is available under each package's `dist` subfolder, so e.g. `./node_modules/@material/button/dist/button.min.css` if you installed it via npm.

It looks like there are some python bindings for libsass as well; here's [one stackoverflow answer](https://stackoverflow.com/a/14413794) about how to integrate Sass specifically with Flask.

**Comment by cubetastic33** [comment]
@kfranqueiro Thank you very much, that cleared a lot of things up. But still, I have a few questions: For compiling SASS to CSS, is it necessary to have Flask compile it for me, or is it fine if I use something like Koala? And, then, you said precompiled JS is available under each package's dist subfolder, but there are only CSS files there!

**Comment by kfranqueiro** [comment]
Some packages, such as MDC Button, don't contain any JavaScript logic, so that might be why you only saw CSS. If you look at e.g. MDC Ripple (which can be used on buttons), you'll see a min.js file as well.

It sounds like Koala is capable of watching for file changes and automatically recompiling, which is really useful for development. Whether you use a GUI or command-line tool for Sass compilation shouldn't matter; it's typical to also want to have a workflow that supports building/deploying straight from a command line, though. Just make sure you can configure it to add the `node_modules` folder containing MDC packages to the Sass include path, like we do in the Getting Started guide for configuring Sass through webpack.

**Comment by cubetastic33** [comment]
@kfranqueiro Thanks a lot! Right now, for the node modules, what I'm doing is, I have a Node JS application, so I install the package there via NPM, and then I just copy the "@material" directory to the CSS folder in my Flask application, and it seem to work very well. So right now, for me, the only problem seems to be compiling ES6. Right now, I'm using the unpkg CDN for JavaScript, but I think I'm missing out on some stuff. Also, the MDC web docs as well as some components (like bottom app bar) still seem to be in development, and lots of the docs don't include what to do if I'm using unpkg JS, and what to include in SASS. I'm able to figure out what to do in SASS pretty easily, just by looking in the folders and finding the .scss files, but figuring out the JS code seems to be a little bit hard!

**Comment by rishabh3112** [comment]
@cubetastic33 if you are using Node then i think you should build templates that you would be serving in flask on node itself and copy only bundled css and js to your statics folder and then use jinja in html files to your as per requirements.

**Comment by rishabh3112** [comment]
You needn't copy whole app to node what i said is that for just templates use node.
You may get example here https://github.com/rishabh3112/rishabh3112.github.io
Where i created index.html with node but used only bundled css and js to serve from master branch

**Comment by kfranqueiro** [comment]
It would be a good idea to separate those, otherwise this issue is going to become incredibly hard to follow. However, I think I can probably address at least one of the two briefly:

The tabs selection looks like the tabs weren't in document flow at the time the component was instantiated. It looks like only the first one is active, but the indicator is the wrong size. Make sure your tabs are actually visible in the DOM before instantiating the component in JS. We're currently redesigning tabs as well, so it's possible the new version may be more resilient to this kind of issue.

For the select, I suspect maybe your markup isn't quite right, but a Codepen example is worth a thousand screenshots. We have a [starter template](https://codepen.io/williamernest/pen/deyxyp) that you can fork and add MDC Web markup to in order to demonstrate issues.

**Comment by kfranqueiro** [comment]
I tried to reproduce your tab-bar-inside-dialog problem but it ended up working anyway: https://codepen.io/kfranqueiro/pen/ZodZVP

However, as a generalized solution, you should be able to call `layout()` on the tab bar component instance after showing the dialog to tell it to recompute layout-dependent styles.

Since it seems you've gotten past your original issue, I'm going to close this. Feel free to open another issue if you encounter something you think is a MDC Web bug; for smaller questions, the web channel on the [Discord](https://discord.gg/material-components) server may be a good idea.
